src/libostree/ostree-repo.c

```diff
@@ -156,6 +156,513 @@ G_DEFINE_TYPE (OstreeRepo, ostree_repo, G_TYPE_OBJECT)
+
+  if (cur_state == LOCK_EX)
+    {
+      g_debug ("Repo already locked exclusively, extending stack");
+      g_queue_push_head (lock->stack, GINT_TO_POINTER (LOCK_EX));
+    }
+  else
+    {
+      int next_state = (flags & LOCK_EX) ? LOCK_EX : LOCK_SH;
+      const char *next_state_name = (flags & LOCK_EX) ? "exclusive" : "shared";
+
+      g_debug ("Locking repo %s", next_state_name);
+      if (!do_repo_lock (lock->fd, flags))
+        return glnx_throw_errno_prefix (error, "Locking repo %s failed",
+                                        next_state_name);
+
+      g_queue_push_head (lock->stack, GINT_TO_POINTER (next_state));
+    }
+
+  return TRUE;
+}
+
+static gboolean
+pop_repo_lock (OstreeRepo  *self,
+               gboolean     blocking,
+               GError     **error)
+{
+  int flags = blocking ? 0 : LOCK_NB;
+
+  GHashTable *lock_table = g_private_get (&repo_lock_table);
+  g_return_val_if_fail (lock_table != NULL, FALSE);
+
+  OstreeRepoLock *lock = g_hash_table_lookup (lock_table, self);
+  g_return_val_if_fail (lock != NULL, FALSE);
+  g_return_val_if_fail (lock->stack != NULL, FALSE);
+  g_return_val_if_fail (lock->fd != -1, FALSE);
+
+  guint stack_len = g_queue_get_length (lock->stack);
+  g_return_val_if_fail (stack_len > 0, FALSE);
+
+  int cur_state = GPOINTER_TO_INT (g_queue_peek_head (lock->stack));
+  g_debug ("Pop lock: state=%s, depth=%u",
+           (cur_state == LOCK_EX) ? "exclusive" : "shared",
+           stack_len);
+  if (stack_len > 1)
+    {
+      int next_state = GPOINTER_TO_INT (g_queue_peek_nth (lock->stack, 1));
+
+      /* Drop back to the previous lock state if it differs */
+      if (next_state != cur_state)
+        {
+          /* We should never drop from shared to exclusive */
+          g_return_val_if_fail (next_state == LOCK_SH, FALSE);
+          g_debug ("Returning lock state to shared");
+          if (!do_repo_lock (lock->fd, next_state | flags))
+            return glnx_throw_errno_prefix (error,
+                                            "Setting repo lock to shared failed");
+        }
+      else
+        g_debug ("Maintaining lock state as %s",
+                 (next_state == LOCK_EX) ? "exclusive" : "shared");
+    }
+  else
+    {
+      /* Lock stack will be empty, unlock */
+      g_debug ("Unlocking repo");
+      if (!do_repo_unlock (lock->fd, flags))
+        return glnx_throw_errno_prefix (error, "Unlocking repo failed");
+    }
+
+  g_queue_pop_head (lock->stack);
+
+  return TRUE;
+}
+
+/**
+ * ostree_repo_lock_push
+ * @self: a #OstreeRepo
+ * @lock_type: the type of lock to acquire
+ * @cancellable: a #GCancellable
+ * @error: a #GError
+ *
+ * Takes a lock on the repository and adds it to the lock stack. If @lock_type
+ * is %OSTREE_REPO_LOCK_SHARED, a shared lock is taken. If @lock_type is
+ * %OSTREE_REPO_LOCK_EXCLUSIVE, an exclusive lock is taken. shared lock is
+ * used. The actual lock state is only changed when locking a previously
```

Comment:
Something looks wrong with the ‘shared lock is used’ mini-sentence. What’s it supposed to mean?

Reply:
Whoops. Probably a rebase mishap. I'll clean that up.

Reply:
Fixed in #1343.